Source/Scene/ModelExperimental/SceneMode2DPipelineStage.js

```diff
@@ -0,0 +1,284 @@
+SceneMode2DPipelineStage.name = "SceneMode2DPipelineStage"; // Helps with debugging
+
+/**
+ * This pipeline stage processes the position attribute of a primitive and adds the relevant
+ * define and uniform matrix to the shader. It also generates new resources for the primitive
+ * in 2D. These resources persist in the runtime primitive so that the typed array used to
+ * store the positional data can be freed.
+ *
+ * This stage must go before the GeometryPipelineStage in the primitive pipeline.
+ *
+ * Processes a primitive. This stage modifies the following parts of the render resources:
+ * <ul>
+ *  <li> creates a vertex buffer for the positions of the primitive projected to 2D
+ *  <li> creates the bounding sphere for the primitive in 2D
+ *  <li> adds a flag to the shader to use 2D positions
+ *  <li> adds a uniform for the view model matrix in 2D
+ * </ul>
+ *
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive.
+ * @param {ModelComponents.Primitive} primitive The primitive.
+ * @param {FrameState} frameState The frame state.
+ *
+ * @private
+ */
+
+SceneMode2DPipelineStage.process = function (
+  renderResources,
+  primitive,
+  frameState
+) {
+  const shaderBuilder = renderResources.shaderBuilder;
+  const runtimePrimitive = renderResources.runtimePrimitive;
+
+  const model = renderResources.model;
+  const modelMatrix = model.sceneGraph.computedModelMatrix;
+  const nodeComputedTransform = renderResources.runtimeNode.computedTransform;
+  const computedModelMatrix = Matrix4.multiplyTransformation(
+    modelMatrix,
+    nodeComputedTransform,
+    scratchModelMatrix
+  );
+
+  const boundingSphere2D = computeBoundingSphere2D(
+    renderResources,
+    computedModelMatrix,
+    frameState
+  );
+
+  runtimePrimitive.boundingSphere2D = boundingSphere2D;
+
+  const positionAttribute = ModelExperimentalUtility.getAttributeBySemantic(
+    primitive,
+    VertexAttributeSemantic.POSITION
+  );
+
+  // If the typed array of the position attribute exists, then
+  // the positions haven't been projected to 2D yet.
+  if (defined(positionAttribute.typedArray)) {
+    const buffer2D = createPositionBufferFor2D(
+      positionAttribute,
+      computedModelMatrix,
+      boundingSphere2D,
+      frameState
+    );
+
+    runtimePrimitive.positionBuffer2D = buffer2D;
+    model._modelResources.push(buffer2D);
+
+    // Unload the typed array
+    positionAttribute.typedArray = undefined;
+  }
+
+  shaderBuilder.addDefine(
+    "USE_2D_POSITIONS",
+    undefined,
+    ShaderDestination.VERTEX
+  );
+
+  shaderBuilder.addUniform("mat4", "u_modelView2D", ShaderDestination.VERTEX);
+
+  const modelMatrix2D = Matrix4.fromTranslation(
+    boundingSphere2D.center,
+    new Matrix4()
+  );
+  const modelView = new Matrix4();
+
+  const camera = frameState.camera;
+  const uniformMap = {
+    u_modelView2D: function () {
+      return Matrix4.multiplyTransformation(
+        camera.viewMatrix,
```

Comment:
Should this be using `context.uniformState.view` instead? or was this intentional?